pkg/flowaggregator/flowaggregator_test.go

```diff
@@ -278,3 +286,86 @@ func createElement(name string, enterpriseID uint32) ipfixentities.InfoElementWi
 }
+
+func TestFlowAggregator_watchConfiguration(t *testing.T) {
+	dbEP := &clickhouseclient.ClickHouseExportProcess{}
+	testcases := []struct {
+		FlowCollectorEnable       bool
+		FlowCollectorAddress      string
+		ClickHouseEnable          bool
+		externalFlowCollectorAddr string
+		message                   []updateParam
+		dbExportProcess           *clickhouseclient.ClickHouseExportProcess
+	}{
+		{true, "10.10.10.10:155", false, "", []updateParam{updateExternalFlowCollectorAddr}, nil},
+		{false, "", true, "addr", []updateParam{disableFlowCollector}, nil},
+		{false, "", true, "", []updateParam{enableClickHouse}, nil},
+		{false, "", true, "", []updateParam{updateClickHouseParam}, dbEP},
+		{true, "10.10.10.10:155", false, "", []updateParam{updateExternalFlowCollectorAddr, disableClickHouse}, dbEP},
+	}
+
+	for i, tc := range testcases {
+		t.Run(fmt.Sprintf("subtest: %d", i), func(t *testing.T) {
+			stopCh := make(chan struct{})
+			opt := util.Options{
+				Config: &flowaggregatorconfig.FlowAggregatorConfig{
+					FlowCollector: flowaggregatorconfig.FlowCollectorConfig{
+						Enable:  tc.FlowCollectorEnable,
+						Address: tc.FlowCollectorAddress,
+					},
+					ClickHouse: flowaggregatorconfig.ClickHouseConfig{
+						Enable: tc.ClickHouseEnable,
+					},
+				},
+			}
+			// create watcher
+			fileName := fmt.Sprintf("test_%d.config", i)
+			configWatcher, err := fsnotify.NewWatcher()
+			assert.NoError(t, err)
+			flowAggregator := &flowAggregator{
+				// use a larger buffer to prevent the buffered channel from blocking
+				updateCh:                  make(chan updateMsg, 100),
+				externalFlowCollectorAddr: tc.externalFlowCollectorAddr,
+				dbExportProcess:           tc.dbExportProcess,
+				configFile:                fileName,
+				configWatcher:             configWatcher,
+			}
+			dir := filepath.Dir(fileName)
+			// create default file
+			f, err := os.Create(flowAggregator.configFile)
+			assert.NoError(t, err)
+			// add default file to watcher
+			err = flowAggregator.configWatcher.Add(dir)
+			assert.NoError(t, err)
+			// chmod and delete file, then create a new one (mimic replacement)
+			err = f.Chmod(0777)
+			assert.NoError(t, err)
+			err = f.Close()
+			assert.NoError(t, err)
+			err = os.Remove(flowAggregator.configFile)
+			assert.NoError(t, err)
+			f, err = os.Create(flowAggregator.configFile)
+			assert.NoError(t, err)
+			b, err := yaml.Marshal(opt.Config)
+			assert.NoError(t, err)
+			_, err = f.Write(b)
+			assert.NoError(t, err)
+			err = f.Close()
+			assert.NoError(t, err)
+			go flowAggregator.watchConfiguration(stopCh)
+
+			for _, message := range tc.message {
+				select {
+				case msg := <-flowAggregator.updateCh:
+					assert.Equal(t, message, msg.param)
+				case <-time.After(1 * time.Minute):
+					assert.NoError(t, fmt.Errorf("timeout"))
+				}
+			}
+			close(stopCh)
+			// wait for watchConfiguration to return to prevent "no such file" error.
+			time.Sleep(5 * time.Second)
```

Comment:
this is much too long a sleep for a unit test

since you are running `watchConfiguration` yourself in this test, it's easy for it to return by using a wait group or a channel

Reply:
I used a channel to close `watchConfiguration `. But when two events happen (stopCH, file event), it will pick a random one to perform. I've added a wg to `watchConfiguration ` to wait util it returns and removed the time.sleep().
Thanks for the suggestion.